test/irmin-pack/test_upgrade.ml

```diff
@@ -0,0 +1,445 @@
+
+    let add_index t =
+      if I.indexing_strategy = "always" then (
+        add_entry_index t "95cb0bfd2977c761298d9624e4b4d4c72a39974a"
+          (133, 23, 'B');
+        add_entry_index t "d9962241e991b771359d8278e79004a334088fde"
+          (156, 44, 'R'))
+      else ();
+      add_entry_index t "8aa02b8a95077bffe8672222ef69c8967e23ed49" (200, 42, 'D')
+
+    let add_dict t = Hashtbl.add t.dict "baba" 2
+
+    let add t =
+      add_suffix t;
+      add_index t;
+      add_dict t
+  end
+
+  module Store = struct
+    module S = struct
+      module Maker = Irmin_pack_unix.Maker (Conf)
+      include Maker.Make (Schema)
+    end
+
+    type repo = S.repo
+
+    let config ?(readonly = false) ?(fresh = true) root =
+      let module Index = Irmin_pack.Indexing_strategy in
+      let indexing_strategy =
+        if I.indexing_strategy = "always" then Index.always else Index.minimal
+      in
+      Irmin_pack.config ~readonly ~indexing_strategy ~lru_size:0 ~fresh root
+
+    let v ~readonly ~fresh root = S.Repo.v (config ~readonly ~fresh root)
+    let close = S.Repo.close
+    let reload = S.reload
+
+    let preload_commit repo =
+      let tree = S.Tree.empty () in
+      let* tree = S.Tree.add tree [ "abba"; "abab" ] "x" in
+      let* commit = S.Commit.v repo ~info:S.Info.empty ~parents:[] tree in
+      let commit_key = S.Commit.key commit in
+      Lwt.return commit_key
+
+    let commit repo =
+      let* o =
+        "94919d4ecc64db71f19ea134e0eedf675512453c"
+        |> hash_of_string
+        |> S.Commit.of_hash repo
+      in
+      match o with
+      | None -> assert false
+      | Some commit ->
+          let tree = S.Commit.tree commit in
+          let* tree = S.Tree.add tree [ "baba" ] "y" in
+          S.Commit.v repo ~info:S.Info.empty ~parents:[] tree
+
+    let pp_hash = Irmin.Type.pp Schema.Hash.t
+
+    let decode_entry_prefix ~off fm =
+      let module CA = S.X.Contents.CA in
+      let entry = CA.read_and_decode_entry_prefix ~off fm in
+      let len = CA.Entry_prefix.total_entry_length entry |> Option.get in
+      (entry.hash, len)
+
+    let dict_find (repo : S.repo) step =
+      let dict = repo.dict in
+      S.Dict.index dict step |> Option.get
+
+    let index_find (repo : S.repo) hash =
+      let index = S.File_manager.index repo.fm in
+      S.Index.find index hash |> Option.get
+
+    let suffix_find (repo : S.repo) off = decode_entry_prefix ~off repo.fm
+
+    let index_iter (repo : S.repo) f =
+      let index = S.File_manager.index repo.fm in
+      S.Index.iter f index
+  end
+
+  exception Skip_the_rest_of_that_test
+
+  let fail_and_skip root error =
+    let* () =
+      Alcotest.check_raises_lwt "open empty/V2 store in RO"
+        (Irmin_pack_unix.Errors.Pack_error error) (fun () ->
+          let* repo = Store.v ~readonly:true ~fresh:false root in
+          Store.close repo)
+    in
+    raise Skip_the_rest_of_that_test
+
+  let test_dict repo model =
+    Hashtbl.iter
```

Comment:
Can you add a comment that says we are only checking for injection and not bijection?

Same for suffix